vcd/datasource_vcd_storage_profile_test.go

```diff
@@ -0,0 +1,140 @@
+		ProviderFactories: testAccProviders,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: configText,
+				Check: resource.ComposeTestCheckFunc(
+					// Ensure that ID matches storage profile URN format (e.g. urn:vcloud:vdcstorageProfile:db4aaa49-7593-4416-93df-37235a1a2c68)
+					resource.TestMatchResourceAttr("data.vcd_storage_profile.sp", "id", regexp.MustCompile(`^urn:vcloud:vdcstorageProfile:`)),
+					resource.TestCheckResourceAttr("data.vcd_storage_profile.sp", "id", storageProfileId),
+					checkStorageProfileOriginatesInParentVdc("data.vcd_storage_profile.sp",
+						params["StorageProfileName"].(string),
+						params["Org"].(string),
+						params["Vdc"].(string)),
+				),
+			},
+		},
+	})
+}
+
+// checkStorageProfileOriginatesInParentVdc tries to evaluate reverse order and ensure that the found storage profile ID
+// does really belong to Org and Vdc specified in datasource
+func checkStorageProfileOriginatesInParentVdc(resource, storageProfileName, orgName, vdcName string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[resource]
+		if !ok {
+			return fmt.Errorf("resource not found: %s", resource)
+		}
+		// Lookup ID field of resource
+		resourceId := rs.Primary.ID
+
+		vcdClient := createTemporaryVCDConnection()
+		adminOrg, err := vcdClient.GetAdminOrgByName(orgName)
+		if err != nil {
+			return fmt.Errorf("error getting adminOrg: %s", err)
+		}
+
+		// Retrieve VDCs
+		allVdcs, err := adminOrg.GetAllVDCs(false)
+		if err != nil {
+			return fmt.Errorf("error getting adminOrg: %s", err)
+		}
+
+		// Check if in any of Orgs child VDCs there is a storage profile that would match ID, Name and Vdc name
+		for _, vdc := range allVdcs {
+			for _, storageProfile := range vdc.Vdc.VdcStorageProfiles.VdcStorageProfile {
+				if storageProfile.ID == resourceId && storageProfile.Name == storageProfileName && vdc.Vdc.Name == vdcName {
+					return nil
+				}
+			}
+		}
+
+		return fmt.Errorf("could not validate storage profile '%s' with ID '%s' belongs to Vdc '%s",
+			storageProfileName, resourceId, testConfig.VCD.Vdc)
+	}
+}
+
+// findStorageProfileIdInVdc should find storage profile ID using the ID that comes from data source
+func findStorageProfileIdInVdc(t *testing.T, storageProfileName string) string {
+	vcdClient := createTemporaryVCDConnection()
+	adminOrg, err := vcdClient.GetAdminOrgByName(testConfig.VCD.Org)
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
+
+	// get all child VDCs
+	childVdcs, err := adminOrg.GetAllVDCs(false)
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
+
+	// Find VDC which should own storage profile defined in test configuration
+	var foundVdc *govcd.Vdc
+	for _, vdc := range childVdcs {
+		// We are looking in correct VDC
+		if vdc.Vdc.Name == testConfig.VCD.Vdc {
+			foundVdc = vdc
+
+		}
+	}
+	if foundVdc == nil {
+		t.Errorf("unable to find Vdc '%s'", testConfig.VCD.Vdc)
```

Comment:
Let's use capital "VDC" in user-facing messages:
```suggestion
		t.Errorf("unable to find VDC '%s'", testConfig.VCD.Vdc)
```

Reply:
Done.